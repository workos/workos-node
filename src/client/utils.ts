import qs from 'qs';

/**
 * Converts an options object to a query string using qs library.
 * Maintains backwards compatibility with existing URL formatting.
 */
export const toQueryString = (
  options: Record<
    string,
    string | string[] | Record<string, string | boolean | number> | undefined
  >,
): string => {
  return qs.stringify(options, {
    arrayFormat: 'repeat',
    // sorts the keys alphabetically to maintain backwards compatibility
    sort: (a, b) => a.localeCompare(b),
    // encodes space as + instead of %20 to maintain backwards compatibility
    format: 'RFC1738',
  });
};
