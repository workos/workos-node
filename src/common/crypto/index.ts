export * from './NodeCryptoProvider';
export * from './SubtleCryptoProvider';
export * from './CryptoProvider';
