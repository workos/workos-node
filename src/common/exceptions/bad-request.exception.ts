import { RequestException } from '../interfaces/request-exception.interface';

export class BadRequestException extends Error implements RequestException {
  readonly status = 400;
  readonly name = 'BadRequestException';
  readonly message: string = 'Bad request';
  readonly code?: string;
  readonly errors?: unknown[];
  readonly requestID: string;

  constructor({
    code,
    errors,
    message,
    requestID,
  }: {
    code?: string;
    errors?: unknown[];
    message?: string;
    requestID: string;
  }) {
    super();

    this.requestID = requestID;

    if (message) {
      this.message = message;
    }

    if (code) {
      this.code = code;
    }

    if (errors) {
      this.errors = errors;
    }
  }
}
