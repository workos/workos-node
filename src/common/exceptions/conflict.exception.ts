import { RequestException } from '../interfaces/request-exception.interface';

export class ConflictException extends Error implements RequestException {
  readonly status = 409;
  readonly name = 'ConflictException';
  readonly requestID: string;

  constructor({
    error,
    message,
    requestID,
  }: {
    error?: string;
    message?: string;
    requestID: string;
  }) {
    super();

    this.requestID = requestID;

    if (message) {
      this.message = message;
    } else if (error) {
      this.message = `Error: ${error}`;
    } else {
      this.message = `An conflict has occurred on the server.`;
    }
  }
}
