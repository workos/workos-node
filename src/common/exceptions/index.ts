export * from './generic-server.exception';
export * from './no-api-key-provided.exception';
export * from './not-found.exception';
export * from './oauth.exception';
export * from './signature-verification.exception';
export * from './unauthorized.exception';
export * from './unprocessable-entity.exception';
