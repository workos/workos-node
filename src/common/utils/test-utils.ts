import fetch, { MockParams } from 'jest-fetch-mock';

export function fetchOnce(
  response: {} = {},
  { status = 200, headers, ...rest }: MockParams = {},
) {
  return fetch.once(JSON.stringify(response), {
    status,
    headers: { 'content-type': 'application/json;charset=UTF-8', ...headers },
    ...rest,
  });
}

export function fetchURL() {
  return fetch.mock.calls[0][0];
}

export function fetchSearchParams() {
  return Object.fromEntries(new URL(String(fetchURL())).searchParams);
}

export function fetchHeaders() {
  return fetch.mock.calls[0][1]?.headers;
}

export function fetchBody({ raw = false } = {}) {
  const body = fetch.mock.calls[0][1]?.body;
  if (body instanceof URLSearchParams) {
    return body.toString();
  }
  if (raw) {
    return body;
  }
  return JSON.parse(String(body));
}
