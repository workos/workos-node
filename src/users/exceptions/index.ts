import { UserResponse } from '../interfaces';

export class UnexpectedUserTypeException extends Error {
  constructor(user: UserResponse) {
    super();
    this.message = `Unsupported user type: ${user.user_type} received from API.`;
  }
}
