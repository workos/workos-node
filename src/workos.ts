import {
  GenericServerException,
  NoApiKeyProvidedException,
  NotFoundException,
  UnauthorizedException,
  UnprocessableEntityException,
  OauthException,
  RateLimitExceededException,
} from './common/exceptions';
import {
  GetOptions,
  PostOptions,
  PutOptions,
  WorkOSOptions,
  WorkOSResponseError,
} from './common/interfaces';
import { DirectorySync } from './directory-sync/directory-sync';
import { Events } from './events/events';
import { Organizations } from './organizations/organizations';
import { OrganizationDomains } from './organization-domains/organization-domains';
import { Passwordless } from './passwordless/passwordless';
import { Portal } from './portal/portal';
import { SSO } from './sso/sso';
import { Webhooks } from './webhooks/webhooks';
import { Mfa } from './mfa/mfa';
import { AuditLogs } from './audit-logs/audit-logs';
import { UserManagement } from './user-management/user-management';
import { BadRequestException } from './common/exceptions/bad-request.exception';
import { FetchClient } from './common/utils/fetch-client';
import { FetchError } from './common/utils/fetch-error';

const VERSION = '7.6.0';

const DEFAULT_HOSTNAME = 'api.workos.com';

export class WorkOS {
  readonly baseURL: string;
  private readonly client: FetchClient;

  readonly auditLogs = new AuditLogs(this);
  readonly directorySync = new DirectorySync(this);
  readonly organizations = new Organizations(this);
  readonly organizationDomains = new OrganizationDomains(this);
  readonly passwordless = new Passwordless(this);
  readonly portal = new Portal(this);
  readonly sso = new SSO(this);
  readonly webhooks = new Webhooks();
  readonly mfa = new Mfa(this);
  readonly events = new Events(this);
  readonly userManagement = new UserManagement(this);

  constructor(readonly key?: string, readonly options: WorkOSOptions = {}) {
    if (!key) {
      // process might be undefined in some environments
      this.key = process?.env.WORKOS_API_KEY;

      if (!this.key) {
        throw new NoApiKeyProvidedException();
      }
    }

    if (this.options.https === undefined) {
      this.options.https = true;
    }

    const protocol: string = this.options.https ? 'https' : 'http';
    const apiHostname: string = this.options.apiHostname || DEFAULT_HOSTNAME;
    const port: number | undefined = this.options.port;
    this.baseURL = `${protocol}://${apiHostname}`;

    if (port) {
      this.baseURL = this.baseURL + `:${port}`;
    }

    let userAgent: string = `workos-node/${VERSION}`;

    if (options.appInfo) {
      const { name, version }: { name: string; version: string } =
        options.appInfo;
      userAgent += ` ${name}: ${version}`;
    }

    this.client = new FetchClient(this.baseURL, {
      ...options.config,
      headers: {
        ...options.config?.headers,
        Authorization: `Bearer ${this.key}`,
        'User-Agent': userAgent,
      },
    });
  }

  get version() {
    return VERSION;
  }

  async post<Result = any, Entity = any>(
    path: string,
    entity: Entity,
    options: PostOptions = {},
  ): Promise<{ data: Result }> {
    const requestHeaders: any = {};

    if (options.idempotencyKey) {
      requestHeaders['Idempotency-Key'] = options.idempotencyKey;
    }

    try {
      return await this.client.post<Entity>(path, entity, {
        params: options.query,
        headers: requestHeaders,
      });
    } catch (error) {
      this.handleFetchError({ path, error });

      throw error;
    }
  }

  async get<Result = any>(
    path: string,
    options: GetOptions = {},
  ): Promise<{ data: Result }> {
    try {
      const { accessToken } = options;
      return await this.client.get(path, {
        params: options.query,
        headers: accessToken
          ? { Authorization: `Bearer ${accessToken}` }
          : undefined,
      });
    } catch (error) {
      this.handleFetchError({ path, error });

      throw error;
    }
  }

  async put<Result = any, Entity = any>(
    path: string,
    entity: Entity,
    options: PutOptions = {},
  ): Promise<{ data: Result }> {
    const requestHeaders: any = {};

    if (options.idempotencyKey) {
      requestHeaders['Idempotency-Key'] = options.idempotencyKey;
    }

    try {
      return await this.client.put<Entity>(path, entity, {
        params: options.query,
        headers: requestHeaders,
      });
    } catch (error) {
      this.handleFetchError({ path, error });

      throw error;
    }
  }

  async delete(path: string, query?: any): Promise<void> {
    try {
      await this.client.delete(path, {
        params: query,
      });
    } catch (error) {
      this.handleFetchError({ path, error });

      throw error;
    }
  }

  emitWarning(warning: string) {
    // process might be undefined in some environments
    if (typeof process?.emitWarning !== 'function') {
      //  tslint:disable:no-console
      return console.warn(`WorkOS: ${warning}`);
    }

    return process.emitWarning(warning, 'WorkOS');
  }

  private handleFetchError({ path, error }: { path: string; error: unknown }) {
    const { response } = error as FetchError<WorkOSResponseError>;

    if (response) {
      const { status, data, headers } = response;
      const requestID = headers.get('X-Request-ID') ?? '';
      const {
        code,
        error_description: errorDescription,
        error,
        errors,
        message,
      } = data;

      switch (status) {
        case 401: {
          throw new UnauthorizedException(requestID);
        }
        case 422: {
          throw new UnprocessableEntityException({
            code,
            errors,
            message,
            requestID,
          });
        }
        case 404: {
          throw new NotFoundException({
            code,
            message,
            path,
            requestID,
          });
        }
        case 429: {
          const retryAfter = headers.get('Retry-After');

          throw new RateLimitExceededException(
            data.message,
            requestID,
            retryAfter ? Number(retryAfter) : null,
          );
        }
        default: {
          if (error || errorDescription) {
            throw new OauthException(
              status,
              requestID,
              error,
              errorDescription,
              data,
            );
          } else if (code && errors) {
            // Note: ideally this should be mapped directly with a `400` status code.
            // However, this would break existing logic for the `OauthException` exception.
            throw new BadRequestException({
              code,
              errors,
              message,
              requestID,
            });
          } else {
            throw new GenericServerException(
              status,
              data.message,
              data,
              requestID,
            );
          }
        }
      }
    }
  }
}
