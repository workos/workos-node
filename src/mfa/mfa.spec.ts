import fetch from 'jest-fetch-mock';
import { fetchOnce, fetchURL, fetchBody } from '../common/utils/test-utils';
import { UnprocessableEntityException } from '../common/exceptions';

import { WorkOS } from '../workos';
import {
  Challenge,
  ChallengeResponse,
  Factor,
  FactorResponse,
  FactorWithSecrets,
  FactorWithSecretsResponse,
  VerifyResponse,
  VerifyResponseResponse,
} from './interfaces';

describe('MFA', () => {
  beforeEach(() => fetch.resetMocks());

  describe('getFactor', () => {
    it('returns the requested factor', async () => {
      const factor: Factor = {
        object: 'authentication_factor',
        id: 'auth_factor_1234',
        createdAt: '2022-03-15T20:39:19.892Z',
        updatedAt: '2022-03-15T20:39:19.892Z',
        type: 'totp',
        totp: {
          issuer: 'WorkOS',
          user: 'some_user',
        },
      };

      const factorResponse: FactorResponse = {
        object: 'authentication_factor',
        id: 'auth_factor_1234',
        created_at: '2022-03-15T20:39:19.892Z',
        updated_at: '2022-03-15T20:39:19.892Z',
        type: 'totp',
        totp: {
          issuer: 'WorkOS',
          user: 'some_user',
        },
      };

      fetchOnce(factorResponse);

      const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU');
      const subject = await workos.mfa.getFactor('test_123');

      expect(subject).toEqual(factor);
    });
  });

  describe('deleteFactor', () => {
    it('sends request to delete a Factor', async () => {
      fetchOnce();
      const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU');

      await workos.mfa.deleteFactor('conn_123');

      expect(fetchURL()).toContain('/auth/factors/conn_123');
    });
  });

  describe('enrollFactor', () => {
    describe('with generic', () => {
      it('enrolls a factor with generic type', async () => {
        const factor: Factor = {
          object: 'authentication_factor',
          id: 'auth_factor_1234',
          createdAt: '2022-03-15T20:39:19.892Z',
          updatedAt: '2022-03-15T20:39:19.892Z',
          type: 'generic_otp',
        };

        const factorResponse: FactorResponse = {
          object: 'authentication_factor',
          id: 'auth_factor_1234',
          created_at: '2022-03-15T20:39:19.892Z',
          updated_at: '2022-03-15T20:39:19.892Z',
          type: 'generic_otp',
        };

        fetchOnce(factorResponse);

        const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU', {
          apiHostname: 'api.workos.dev',
        });

        const subject = await workos.mfa.enrollFactor({
          type: 'generic_otp',
        });

        expect(subject).toEqual(factor);
      });
    });

    describe('with totp', () => {
      it('enrolls a factor with totp type', async () => {
        const factor: FactorWithSecrets = {
          object: 'authentication_factor',
          id: 'auth_factor_1234',
          createdAt: '2022-03-15T20:39:19.892Z',
          updatedAt: '2022-03-15T20:39:19.892Z',
          type: 'totp',
          totp: {
            issuer: 'WorkOS',
            qrCode: 'qr-code-test',
            secret: 'secret-test',
            uri: 'uri-test',
            user: 'some_user',
          },
        };

        const factorResponse: FactorWithSecretsResponse = {
          object: 'authentication_factor',
          id: 'auth_factor_1234',
          created_at: '2022-03-15T20:39:19.892Z',
          updated_at: '2022-03-15T20:39:19.892Z',
          type: 'totp',
          totp: {
            issuer: 'WorkOS',
            qr_code: 'qr-code-test',
            secret: 'secret-test',
            uri: 'uri-test',
            user: 'some_user',
          },
        };

        fetchOnce(factorResponse);
        const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU', {
          apiHostname: 'api.workos.dev',
        });

        const subject = await workos.mfa.enrollFactor({
          type: 'totp',
          issuer: 'WorkOS',
          user: 'some_user',
        });

        expect(subject).toEqual(factor);
      });
    });

    describe('with sms', () => {
      it('enrolls a factor with sms type', async () => {
        const factor: Factor = {
          object: 'authentication_factor',
          id: 'auth_factor_1234',
          createdAt: '2022-03-15T20:39:19.892Z',
          updatedAt: '2022-03-15T20:39:19.892Z',
          type: 'sms',
          sms: {
            phoneNumber: '+15555555555',
          },
        };

        const factorResponse: FactorResponse = {
          object: 'authentication_factor',
          id: 'auth_factor_1234',
          created_at: '2022-03-15T20:39:19.892Z',
          updated_at: '2022-03-15T20:39:19.892Z',
          type: 'sms',
          sms: {
            phone_number: '+15555555555',
          },
        };

        fetchOnce(factorResponse);

        const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU', {
          apiHostname: 'api.workos.dev',
        });

        const subject = await workos.mfa.enrollFactor({
          type: 'sms',
          phoneNumber: '+1555555555',
        });

        expect(subject).toEqual(factor);
      });

      describe('when phone number is invalid', () => {
        it('throws an exception', async () => {
          fetchOnce(
            {
              message: `Phone number is invalid: 'foo'`,
              code: 'invalid_phone_number',
            },
            {
              status: 422,
              headers: { 'X-Request-ID': 'req_123' },
            },
          );

          const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU', {
            apiHostname: 'api.workos.dev',
          });

          await expect(
            workos.mfa.enrollFactor({
              type: 'sms',
              phoneNumber: 'foo',
            }),
          ).rejects.toThrow(UnprocessableEntityException);
        });
      });
    });
  });

  describe('challengeFactor', () => {
    describe('with no sms template', () => {
      it('challenge a factor with no sms template', async () => {
        const challenge: Challenge = {
          object: 'authentication_challenge',
          id: 'auth_challenge_1234',
          createdAt: '2022-03-15T20:39:19.892Z',
          updatedAt: '2022-03-15T20:39:19.892Z',
          expiresAt: '2022-03-15T21:39:19.892Z',
          code: '12345',
          authenticationFactorId: 'auth_factor_1234',
        };

        const challengeResponse: ChallengeResponse = {
          object: 'authentication_challenge',
          id: 'auth_challenge_1234',
          created_at: '2022-03-15T20:39:19.892Z',
          updated_at: '2022-03-15T20:39:19.892Z',
          expires_at: '2022-03-15T21:39:19.892Z',
          code: '12345',
          authentication_factor_id: 'auth_factor_1234',
        };

        fetchOnce(challengeResponse);

        const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU', {
          apiHostname: 'api.workos.dev',
        });

        const subject = await workos.mfa.challengeFactor({
          authenticationFactorId: 'auth_factor_1234',
        });

        expect(subject).toEqual(challenge);
      });
    });

    describe('with sms template', () => {
      it('challenge a factor with sms template', async () => {
        const challenge: Challenge = {
          object: 'authentication_challenge',
          id: 'auth_challenge_1234',
          createdAt: '2022-03-15T20:39:19.892Z',
          updatedAt: '2022-03-15T20:39:19.892Z',
          expiresAt: '2022-03-15T21:39:19.892Z',
          code: '12345',
          authenticationFactorId: 'auth_factor_1234',
        };

        const challengeResponse: ChallengeResponse = {
          object: 'authentication_challenge',
          id: 'auth_challenge_1234',
          created_at: '2022-03-15T20:39:19.892Z',
          updated_at: '2022-03-15T20:39:19.892Z',
          expires_at: '2022-03-15T21:39:19.892Z',
          code: '12345',
          authentication_factor_id: 'auth_factor_1234',
        };

        fetchOnce(challengeResponse);

        const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU', {
          apiHostname: 'api.workos.dev',
        });

        const subject = await workos.mfa.challengeFactor({
          authenticationFactorId: 'auth_factor_1234',
          smsTemplate: 'This is your code: 12345',
        });

        expect(fetchBody()).toEqual({
          sms_template: 'This is your code: 12345',
        });
        expect(subject).toEqual(challenge);
      });
    });
  });

  describe('verifyChallenge', () => {
    describe('verify with successful response', () => {
      it('verifies a successful factor', async () => {
        const verifyResponse: VerifyResponse = {
          challenge: {
            object: 'authentication_challenge',
            id: 'auth_challenge_1234',
            createdAt: '2022-03-15T20:39:19.892Z',
            updatedAt: '2022-03-15T20:39:19.892Z',
            expiresAt: '2022-03-15T21:39:19.892Z',
            code: '12345',
            authenticationFactorId: 'auth_factor_1234',
          },
          valid: true,
        };

        const verifyResponseResponse: VerifyResponseResponse = {
          challenge: {
            object: 'authentication_challenge',
            id: 'auth_challenge_1234',
            created_at: '2022-03-15T20:39:19.892Z',
            updated_at: '2022-03-15T20:39:19.892Z',
            expires_at: '2022-03-15T21:39:19.892Z',
            code: '12345',
            authentication_factor_id: 'auth_factor_1234',
          },
          valid: true,
        };

        fetchOnce(verifyResponseResponse);

        const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU', {
          apiHostname: 'api.workos.dev',
        });

        const subject = await workos.mfa.verifyChallenge({
          authenticationChallengeId: 'auth_challenge_1234',
          code: '12345',
        });

        expect(fetchBody()).toEqual({
          code: '12345',
        });
        expect(subject).toEqual(verifyResponse);
      });
    });

    describe('when the challenge has been previously verified', () => {
      it('throws an exception', async () => {
        fetchOnce(
          {
            message: `The authentication challenge '12345' has already been verified.`,
            code: 'authentication_challenge_previously_verified',
          },
          {
            status: 422,
            headers: { 'X-Request-ID': 'req_123' },
          },
        );

        const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU', {
          apiHostname: 'api.workos.dev',
        });

        await expect(
          workos.mfa.verifyChallenge({
            authenticationChallengeId: 'auth_challenge_1234',
            code: '12345',
          }),
        ).rejects.toThrow(UnprocessableEntityException);
        expect(fetchBody()).toEqual({
          code: '12345',
        });
      });
    });

    describe('when the challenge has expired', () => {
      it('throws an exception', async () => {
        fetchOnce(
          {
            message: `The authentication challenge '12345' has expired.`,
            code: 'authentication_challenge_expired',
          },
          {
            status: 422,
            headers: { 'X-Request-ID': 'req_123' },
          },
        );

        const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU', {
          apiHostname: 'api.workos.dev',
        });

        await expect(
          workos.mfa.verifyChallenge({
            authenticationChallengeId: 'auth_challenge_1234',
            code: '12345',
          }),
        ).rejects.toThrow(UnprocessableEntityException);
        expect(fetchBody()).toEqual({
          code: '12345',
        });
      });

      it('exception has code', async () => {
        fetchOnce(
          {
            message: `The authentication challenge '12345' has expired.`,
            code: 'authentication_challenge_expired',
          },
          {
            status: 422,
            headers: { 'X-Request-ID': 'req_123' },
          },
        );

        const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU', {
          apiHostname: 'api.workos.dev',
        });

        try {
          await workos.mfa.verifyChallenge({
            authenticationChallengeId: 'auth_challenge_1234',
            code: '12345',
          });
        } catch (error) {
          expect(error).toMatchObject({
            code: 'authentication_challenge_expired',
          });
        }
        expect(fetchBody()).toEqual({
          code: '12345',
        });
      });
    });
  });
});
