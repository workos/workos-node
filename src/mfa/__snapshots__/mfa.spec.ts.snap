// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MFA challengeFactor with no sms template enrolls a factor 1`] = `Promise {}`;

exports[`MFA challengeFactor with totp challenge a factor with sms template 1`] = `Promise {}`;

exports[`MFA enrollFactor with generic enrolls a factor with generic type 1`] = `Promise {}`;

exports[`MFA enrollFactor with sms enrolls a factor with sms type 1`] = `Promise {}`;

exports[`MFA enrollFactor with totp enrolls a factor with totp type 1`] = `Promise {}`;

exports[`MFA getFactor throws an error for incomplete arguments 1`] = `Promise {}`;

exports[`MFA verifyFactor verifies a factor 1`] = `Promise {}`;
